tests/lambda/big_step.rs

```diff
@@ -19,35 +19,43 @@ pub struct Candidate {
 }
 
 // applies rewrites (only beta-reduction) for all applicable situations.
-pub fn rewrite_big_step(eg: &mut EGraph<Lambda>) {
-    for cand in candidates(eg) {
-        let app_id = eg.lookup(&cand.app).unwrap();
-
-        let Lambda::App(l, t) = cand.app.clone() else {
-            panic!()
-        };
-        let Lambda::Lam(Bind { slot: x, elem: b }) = cand.lam.clone() else {
-            panic!()
-        };
-        assert_eq!(x, Slot::numeric(0));
-
-        // l.m :: slots(lam) -> slots(app)
-        let mut m = l.m.clone();
-
-        // if x is a public slot of "app", we'd need to rename. But as x should always be s0 this shouldn't come up.
-        assert!(!m.contains_key(x));
-
-        m.insert(x, x);
-
-        let b = b.apply_slotmap(&m);
-
-        let new_id = subst(b, x, t, eg);
-        eg.union_justified(
-            &new_id,
-            &app_id,
-            Some("big-step-beta-reduction".to_string()),
-        );
+pub fn rewrite_big_step() -> Vec<Rewrite<Lambda>> {
+    let searcher = Box::new(|eg: &EGraph<Lambda>| candidates(eg));
+    let applier = Box::new(|cands: Vec<Candidate>, eg: &mut EGraph<Lambda>| {
+        for cand in cands {
+            let app_id = eg.lookup(&cand.app).unwrap();
+
+            let Lambda::App(l, t) = cand.app.clone() else {
+                panic!()
+            };
+            let Lambda::Lam(Bind { slot: x, elem: b }) = cand.lam.clone() else {
+                panic!()
+            };
+            assert_eq!(x, Slot::numeric(0));
+
+            // l.m :: slots(lam) -> slots(app)
+            let mut m = l.m.clone();
+
+            // if x is a public slot of "app", we'd need to rename. But as x should always be s0 this shouldn't come up.
+            assert!(!m.contains_key(x));
+
+            m.insert(x, x);
+
+            let b = b.apply_slotmap(&m);
+
+            let new_id = subst(b, x, t, eg);
+            eg.union_justified(
+                &new_id,
+                &app_id,
+                Some("big-step-beta-reduction".to_string()),
+            );
+        }
+    });
+    vec![RewriteT {
+        searcher: searcher,
+        applier: applier,
     }
+    .into()]
```

Comment:
Hey, small nit: Currently the "searching" part is happening in the applier. Doesn't matter too much for now, but I tihnk the whole loop should be in the searcher. And the searcher just returns a "Vec<(AppliedId, AppliedId)>" that are merges he's prepared.

Reply:
Not a blocker definitely, but the reason I'd prefer this is that the searcher things can run in parallel (in the future), while the applier not. So it's always good to make more stuff in the searcher.
Plus: just storing the final `Vec<(AppliedId, AppliedId)>` should be much less data to transfer to the applier.